<a href="https://colab.research.google.com/github/shashwatshukla123/class-120/blob/main/class_120.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

## Naïve: 
It is called Naïve because it assumes that the occurrence of a certain feature is independent of the occurrence of other features.
\
 Such as if the fruit is identified on the bases of color, shape, and taste, then red, spherical, and sweet fruit is recognized as an apple. Hence each feature individually contributes to identify that it is an apple without depending on each other.
\
It is a superised machine learnig algorithm based on bayes probability threom.
\
Naïve Bayes assume that there is no corelation betwwen the feature in a dataset used to train the model.
\
Bayes threom describe the probility of an event based n prior knewledge or coditions that might be related to the events.
\
example: If the probilty of someone having diabetes is related to his her age , then by useing the bayes threom the age can be used to predict probability of having diabetes.

In [13]:
import random
import plotly.express as px
import plotly.figure_factory as ff
import statistics
import csv
import pandas as pd
import plotly.graph_objects as go
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt  
from sklearn.linear_model import LogisticRegression
from sklearn.preprocessing import StandardScaler
from sklearn.model_selection import train_test_split 
from sklearn.metrics import confusion_matrix
from sklearn.metrics import accuracy_score 
from sklearn.model_selection import train_test_split
from sklearn import metrics
from sklearn.tree import export_graphviz
from sklearn.externals.six import StringIO 
from IPython.display import Image
import pydotplus
from sklearn.model_selection import train_test_split
from sklearn.naive_bayes import GaussianNB

In [6]:
df = pd.read_csv('diabetes2.csv')
print(df.head())

   glucose  bloodpressure  diabetes
0       40             85         0
1       40             92         0
2       45             63         1
3       45             80         0
4       40             73         1


In [21]:
x = df[['bloodpressure','glucose']]
y = df['diabetes']

In [22]:
x_train, x_test, y_train, y_test = train_test_split(x, y, test_size=0.25, random_state=42)

In [25]:
sc = StandardScaler()
x_train = sc.fit_transform(x_train)
x_test = sc.fit_transform(x_test) 
model = GaussianNB()
model.fit(x_train, y_train) 
y_pred = model.predict(x_test) 
accuracy = accuracy_score(y_test, y_pred) 
print(accuracy)

0.9437751004016064


In [26]:
# Apply logistic regression
x = df[['bloodpressure','glucose']]
y = df['diabetes']
x_train_2, x_test_2, y_train_2, y_test_2 = train_test_split(x, y, test_size=0.25, random_state=42)
sc = StandardScaler()
x_train_2 = sc.fit_transform(x_train_2)
x_test_2 = sc.fit_transform(x_test_2) 
model_2 = LogisticRegression(random_state = 0) 
model_2.fit(x_train_2, y_train_2)
y_pred_2 = model_2.predict(x_test_2)
accuracy = accuracy_score(y_test_2, y_pred_2)
print(accuracy)

0.9156626506024096


# Conclusion
While the accuracy score for both the datasets was close, with Naive Bayes giving us an accuracy of 94.4% and logistic regression giving us an accuracy of 91.6%, Naive Bayes still performed better. The reason for this is that if we look at our features again, we can see that the Glucose and the Blood Pressure had no correlation with each other. They both contributed individually to whether a person would have diabetes or not. This is exactly what Naive Bayes algorithm assumes, that all the features contribute individually to the outcome. This was for the case of where Naive Bayes outperforms Logistic Regression, but let's see an example of the case where Logistic Regression outperforms Naive Bayes.

In [29]:
df = pd.read_csv('income.csv')
print(df.head())
print(df.describe())

   age          workclass  ...  native-country  income
0   39          State-gov  ...   United-States   <=50K
1   50   Self-emp-not-inc  ...   United-States   <=50K
2   38            Private  ...   United-States   <=50K
3   53            Private  ...   United-States   <=50K
4   28            Private  ...            Cuba   <=50K

[5 rows x 14 columns]
                age  education-num  capital-gain  capital-loss  hours-per-week
count  45222.000000   45222.000000  45222.000000  45222.000000    45222.000000
mean      38.547941      10.118460   1101.430344     88.595418       40.938017
std       13.217870       2.552881   7506.430084    404.956092       12.007508
min       17.000000       1.000000      0.000000      0.000000        1.000000
25%       28.000000       9.000000      0.000000      0.000000       40.000000
50%       37.000000      10.000000      0.000000      0.000000       40.000000
75%       47.000000      13.000000      0.000000      0.000000       45.000000
max       90.00

From the given data, we will consider the following fields to determine the salary of a person - Age 
\
Hours Per Week
\
 Education Number 
 \
 Capital Gain 
 \
 Capital Loss

In [32]:
X = df[["age", "hours-per-week", "education-num", "capital-gain", "capital-loss"]] 
y = df["income"]
x_train_1, x_test_1, y_train_1, y_test_1 = train_test_split(X, y, test_size=0.25, random_state=42)

In [34]:
sc = StandardScaler()
x_train_1 = sc.fit_transform(x_train_1)
x_test_1 = sc.fit_transform(x_test_1) 
model = GaussianNB()
model.fit(x_train_1, y_train_1) 
y_pred_1 = model.predict(x_test_1) 
accuracy = accuracy_score(y_test_1, y_pred_1) 
print(accuracy)

0.7896692021935255


In [35]:
# Apply logistic regression
x = df[["age", "hours-per-week", "education-num", "capital-gain", "capital-loss"]]
y = df['income']
x_train_2, x_test_2, y_train_2, y_test_2 = train_test_split(x, y, test_size=0.25, random_state=42)
sc = StandardScaler()
x_train_2 = sc.fit_transform(x_train_2)
x_test_2 = sc.fit_transform(x_test_2) 
model_2 = LogisticRegression(random_state = 0) 
model_2.fit(x_train_2, y_train_2)
y_pred_2 = model_2.predict(x_test_2)
accuracy = accuracy_score(y_test_2, y_pred_2)
print(accuracy)

0.8116929064213692


While the accuracy score for both the datasets was close, with Naive Bayes giving us an accuracy of 78.9% and logistic regression giving us an accuracy of 81.1%, Logistic regression still performed better. 
\
The reason is that in this dataset, not all features contribute individually to the outcome. For example, there have been people of all age groups earning both less than and more than 50K. There have also been people with all education numbers that have an income of both less and more than 50K. Here, the combination of all the features is a better predictor of whether a person is earning more than or less than 50K, instead of all features having their individual contribution.

# Conclusion
IF they are co-related use logistic regression, if not use Naïve Bayes for better predicuion.